File: importers/utils/import_helpers/string_matcher.py
import re
import os.path


def load_lines_from_file(filename):			
	with open(filename, 'r') as f:
		return f.readlines()


def single_space(s):
	while True:
		len1 = len(s)
		s = s.replace("  ", " ")
		if len(s) == len1:
			break

	return s


def strip_all(list1):
	return [single_space(s.strip().lower()) for s in list1]


def sanitize_regex(regex1):
	regex1 = regex1.strip('\n').strip('\r').lower()
	if '$' in regex1:
		regex1 = regex1.rstrip()
	if '^' in regex1:
		regex1 = regex1.lstrip()
	
	return regex1

	
def sanitize_all_regex(list1):
	return [sanitize_regex(regex1) for regex1 in list1]


def regex_matches_any(s, regex_list):
	for regex in regex_list:
		p = re.compile(regex)
		if p.search(s):
			return True
	
	return False


class StringMatcher():
	def __init__(self, path_prefix):
		filename = path_prefix + '_names.txt'
		self.names = []
		self.not_names = []
		self.name_regex = []
		self.not_name_regex = []
		if os.path.isfile(filename):
			self.names = strip_all(load_lines_from_file(filename))
		
		filename = path_prefix + '_name_regex.txt'
		if os.path.isfile(filename):
			self.name_regex = sanitize_all_regex(load_lines_from_file(filename))

		filename = path_prefix + '_not_name_regex.txt'
		if os.path.isfile(filename):
			self.not_name_regex = sanitize_all_regex(load_lines_from_file(filename))

		filename = path_prefix + '_not_names.txt'
		if os.path.isfile(filename):
			self.not_names = strip_all(load_lines_from_file(filename))

	def applies_to_name(self, location_name):
		location_name = single_space(location_name.strip().lower())
		if location_name in self.not_names:
			return False

		if location_name in self.names:
			return True

		if self.not_name_regex:
			if regex_matches_any(location_name, self.not_name_regex):
				return False

		if self.name_regex:
			if regex_matches_any(location_name, self.name_regex):
				return True

		return False

File: app/app/Libraries/StringMatcher.php
<?php

namespace App\Libraries;

/*
    StringMatcher is very similar to
    importers/utils/import_helpers/string_matcher.py
    but written in PHP
*/

class StringMatcher
{
    public function __construct($path_prefix)
    {
        $this->names = [];
        $this->not_names = [];
        $this->name_regex = [];
        $this->not_name_regex = [];
        $filename = $path_prefix . '_names.txt';
        if (file_exists($filename)) {
            $this->names = StringMatcher::loadLinesFromFile($filename);
        }
        $filename = $path_prefix . '_name_regex.txt';
        if (file_exists($filename)) {
            $this->name_regex = StringMatcher::sanitizeAllRegex(StringMatcher::loadLinesFromFile($filename));
        }
        $filename = $path_prefix . '_not_name_regex.txt';
        if (file_exists($filename)) {
            $this->not_name_regex = StringMatcher::sanitizeAllRegex(StringMatcher::loadLinesFromFile($filename));
        }
        $filename = $path_prefix . '_not_names.txt';
        if (file_exists($filename)) {
            $this->not_names = StringMatcher::stripAll(StringMatcher::loadLinesFromFile($filename));
        }
    }

    private static function loadLinesFromFile($filename)
    {
        return explode("\n", file_get_contents($filename));
    }

    public static function singleSpace($s)
    {
        return preg_replace('/\s+/', ' ', $s);
    }

    private static function stripAll($list1)
    {
        return array_map(function ($s) {
                return strtolower(StringMatcher::singleSpace($s));
        },
            $list1);
    }

    public static function sanitizeRegex($regex1)
    {
        $regex1 = strtolower(preg_replace("/\r|\n/", "", $regex1));
        if (strpos($regex1, '$') !== false) {
            $regex1 = rtrim($regex1);
        }
        if (strpos($regex1, '^') !== false) {
            $regex1 = ltrim($regex1);
        }

        return '/'.$regex1.'/';
    }
        
    private static function sanitizeAllRegex($regexArray1)
    {
        return array_map('self::sanitizeRegex', $regexArray1);
    }

    private static function regexMatchesAny($s, $regexArray)
    {
        $result = false;
        foreach ($regexArray as $reg) {
            if (preg_match($reg, $s)) {
                $result = true;
            }
        }
        return $result;
    }
    
    public function appliesToName($location_name)
    {
        $location_name = StringMatcher::singleSpace(strtolower(trim($location_name)));
        if (in_array($location_name, $this->not_names)) {
            return false;
        }

        if (in_array($location_name, $this->names)) {
            return true;
        }

        if ($this->not_name_regex) {
            if (StringMatcher::regexMatchesAny($location_name, $this->not_name_regex)) {
                return false;
            }
        }

        if ($this->name_regex) {
            if (StringMatcher::regexMatchesAny($location_name, $this->name_regex)) {
                return true;
            }
        }

        return false;
    }
}
